### ISSUE TITLE
Using header value for authentication

### ISSUE THREAD

**Issue by icsy7867** [uxbox/uxbox#1766]
**Is your feature request related to a problem? Please describe.**
For many of my docker projects, I use an apache container running mellon to provide SAML authentication from our ADFS servers.  The implementation is a reverse proxy that then connects to a web frontend. It works great, however integrating with other services can be challenging.  OpenID should work, however there are a lot of other reverse proxy authentication solutions that can utilize this as well.

**Describe the solution you'd like**
I recently deployed netbox, which is a great opensource DCIM project.  They have implemented this in a pretty well thought out way.
https://netbox.readthedocs.io/en/stable/configuration/remote-authentication/
using environment variables you can set the header value to read from.

**Describe alternatives you've considered**
Will probably use openID instead.  However I thought this would be fairly easy to implement overall.  However not sure if there are enough users who would see the benefit.

**Comment by niwinz** [comment]
Hello @icsy7867
I'm not clearly understand the feature-request/enhacement. Do you suggest implement SAML?

**Comment by icsy7867** [comment]
> Hello @icsy7867 I'm not clearly understand the feature-request/enhacement. Do you suggest implement SAML?

Yes and no.  Adding saml, openid or others is always nice.  However I'm specifically asking for header off.  This occurs when someone uses authentication via a reverse proxy.  In my case using Mellon to handle SAML and then proxying to the container/service.  Instead of adding in advanced concepts such as SAML or other plugins/feature, you would simply grab the username/email from a header valid, such as : HTTP_X_REMOTE_USER or HTTP_REMOTE_USER.

The reverse proxy handles the value and it's a great way to use a reverse proxy for authentication and relatively easy to implement.

**Comment by niwinz** [comment]
Oh, I think I understand it now:
Correct me if I'm wrong:, the main idea is perform authentication on the reverse proxy, and the reverse proxy is responsible to set this header for authenticated users?

In this case, what will come in this header: `HTTP_X_REMOTE_USER` ? The profile id UUID, some JWT?  For implement it I will need to know how it exactly works.

**Comment by icsy7867** [comment]
I appreciate the responsiveness!  And yes I think you have it.

For my use case, which uses Apache and Mellon to handle SAML/SSO, I can set whatever header I want.  However there is some concern as nginx does not always respect differently named headers.  I would stick with the standardized "HTTP_REMOTE_USER" or "REMOTE_USER" (as the HTTP_ portion will be appended).

However, at that point I would add a config option, for "PROXY_HEADER_ENABLED" which is a boolean, true or false, and another for "PROXY_HEADER_NAME", where the user can set this to a different value, or it defaults to "REMOTE_USER" or "HTTP_REMOTE_USER".

In netbox, in our environment we are using ADFS, and set the "REMOTE_USER" to the samAccountName/Username.  However when using SAML/SSO you can provide whatever claim you wish (email, SID, etc...).  Ultimately if the REMOTE_USER, this user is created in the database if they dont exist and this value is also used as their username.  

You could also pass additional variables as well.  If the DB wants username and email address, that would be easily done as well using an additional header value.

**Comment by niwinz** [comment]
On penpot, right now all external (OIDC mainly) authentication flows requires the `email` as mandatory field and the `name` (that maybe the full name or first name, and in some extreme cases we can implement some kind of derivation of it from the email).

So, having this in context, I don't kwow how it be better implemented on header parsing:

(1)
```bash
REMOTE_USER="my.email@penpot.app,My Name"
```

(2)
```bash
REMOTE_USER="my.email@penpot.app"
REMOTE_USER_NAME="My Name"
```

And on the backend configuration, it is pretty easy to add a new one specific to this scheme.

An other question (mainly because I really not familiar with this authentication scheme): these headers will come on all requests? or only on the login/register? I mean, I need to understand the data flow, in order to understand how implement this.

On the OIDC, it is easy, the flow is:
1. user clicks button, that makes request to backend, backend generates a redirect uri using
2. redirect user to the generated URI to the external provider,
3. once user authenticates in the external provider, the user is redirected it back to app with some token.
4. the token is exchanged for access-token,  and used this access token to access the basic user info (email, name)
5. login-or-register here, and then create session cookie and set it.
6. the following requests are already authenticated thanks to the cookie.

On the auth scheme you explains, how the data will flow?

**Comment by icsy7867** [comment]
This is one of the neat things about handling the auth via a reverse proxy and header value.  Is the flow doesn't really matter. Simply reading the value in those headers means authentication happened.  A user cannot get through the reverse proxy authentication without successfully authenticating (If that makes sense).

For you first question, I think two header values makes sense.  I am going to send a short example of how i handle this in another php based web app I made.  In this example I am setting X-WEBAUTH-USER and X-WEBAUTH-EMAIL.  Of course your app wouldnt need to do ANY of this, except for reading the header value.  This is all done with the reverse proxy/saml.  This excerpt is from a password/secret text sharer I created for my organization.

```
<VirtualHost _default_:80>
ServerName https://pass.orgranization.com
ProxyPass /mellon/ !
<Location / >

               Require valid-user
                AuthType "Mellon"
                MellonEnable "auth"
                MellonVariable "cookie"
	       MellonSecureCookie on
                MellonCookiePath /
                MellonUser "NAME_ID"
                MellonSessionDump Off
                MellonSamlResponseDump Off
                MellonEndpointPath "/mellon"
                MellonDefaultLoginPath "/"
                MellonSessionLength 28800
                MellonSignatureMethod rsa-sha256

    # service provider metadata, cert, and key
    MellonSPPrivateKeyFile /etc/apache2/mellon/saml_sp.key
    MellonSPCertFile /etc/apache2/mellon/saml_sp.cert
    MellonSPMetadataFile /etc/apache2/mellon/saml_sp.xml
    MellonIdpMetadataFile /etc/apache2/mellon/saml_idp.xml
    ProxyPass "http://pass/"
    ProxyPassReverse "http://pass/"


    RequestHeader set "X-Forwarded-Proto" expr=%{REQUEST_SCHEME}
    RequestHeader set "X-Forwarded-SSL" expr=%{HTTPS}

    MellonSetEnvNoPrefix REMOTE_USER NAME_ID
    MellonSetEnvNoPrefix REMOTE_EMAIL emailaddress
    MellonSetEnvNoPrefix "ADFS_EMAIL" "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/emailaddress"

    RequestHeader set X-WEBAUTH-USER %{REMOTE_USER}e env=REMOTE_USER
    RequestHeader set X-WEBAUTH-EMAIL %{ADFS_EMAIL}e env=ADFS_EMAIL

</Location>
</VirtualHost>
```


And then PHP to retrieve the value from the header:
`
$User = isset($_SERVER['HTTP_X_WEBAUTH_USER']) ? strtolower($_SERVER['HTTP_X_WEBAUTH_USER']) : '';
$Email = isset($_SERVER['HTTP_X_WEBAUTH_EMAIL']) ? strtolower($_SERVER['HTTP_X_WEBAUTH_EMAIL']) : 'empty';
`

I love using reverse proxy/saml sso for authentication as I have a dockerized solution for this.   You can deploy the web service/container without any extra configuration (For the most part) and simply have the reverse proxy exposed externall.

With most reverse proxy authentication mechanisms though, it typically follows this path:

1. User hits reverse proxy front end.  In this case, lets call it https://penpot.organization.com. 
2. Mellon authentication is enabled, and redirects to the iDP/External provider for authentication. 
3. users either successfully authenticates, or fails to do so.  If successful, you receive your SAML/SSO cookie (Or you use an existing SAML session cookie).  If not successful, you get sent to an unauthorized page.
4. Mellon is successful, sets header value and forwards traffic to endpoint (In my case, a penpot docker container on an internal only network).
5. You finally see/hit penpot web app.

The reason I really love this authentication with docker, is the authentication/access component is almost completely separate from the webapp.  With ADFS, you can even control access via an AD group which is handled on the reverse proxy before the actual web app.

Lastly, even if a user fakes a header value, and sends it along, it does not matter, because the headers you send to the reverse proxy are not the same as what the reverse proxy sends to the internal web app.

**Comment by niwinz** [comment]
we don't forget about this, just can't take enough quality time for this for now

**Comment by ashwin153** [comment]
@niwinz I'm also interested in this functionality. Is it possible to have it reprioritized?
